src/realm/object-store/sync/generic_network_transport.hpp

```diff
@@ -234,15 +234,96 @@ struct Response {
     util::Optional<ClientErrorCode> client_error_code;
 };
 
+struct HttpCompletionImpl;
+
+struct HttpCompletion {
+    HttpCompletion() noexcept = default;
+    HttpCompletion(HttpCompletion&& copy) noexcept
+        : impl(std::move(copy.impl))
+    {
+    }
+    HttpCompletion(std::unique_ptr<HttpCompletionImpl>&& impl) noexcept
+        : impl(std::move(impl))
+    {
+    }
+    HttpCompletion& operator=(HttpCompletion&& copy) noexcept
+    {
+        impl = std::move(copy.impl);
+        return *this;
+    }
+
+    operator bool() const noexcept
+    {
+        return impl != nullptr;
+    }
+    void operator()(const Response& response);
+
+    const Request& request() const;
+
+    // Helper functions for the c_api interface
+    inline void* release()
+    {
+        REALM_ASSERT(impl != nullptr);
+        return impl.release();
+    }
+    HttpCompletion(void* context)
+        : impl(static_cast<HttpCompletionImpl*>(context))
+    {
+    }
+
+    // Needed for testing purposes
+    void operator()(Request&& request, const Response& response);
+
+protected:
+    std::unique_ptr<HttpCompletionImpl> impl;
+};
 
-using HttpCompletion = util::UniqueFunction<void(const Request&, const Response&)>;
 
 /// Generic network transport for foreign interfaces.
 struct GenericNetworkTransport {
-    virtual void send_request_to_server(Request&& request, HttpCompletion&& completion_block) = 0;
+    virtual void send_request_to_server(const Request& request, HttpCompletion&& completion) = 0;
```

Comment:
This seems way more complicated than is necessary. I think part of the issue is that you seem to be trying to keep the API for implementors of `GenericNetworkTransport` the same as for consumers. Often it makes sense to have an interface provide different APIs for each side. Why not just do something simple like: 

```cpp
struct GenericNetworkTransport {
    void send_request_to_server(Request&& request, util::UniqueFunction<void(const Request&, const Response&)>&& cb)
    {
        auto request_ptr = std::make_unique<Request>(std::move(request));
        const auto& request_ref = *request_ptr;
        do_send_request_to_server(
            request_ref, [request_ptr = std::move(request_ptr), cb = std::move(cb)](const Response& response) {
                cb(*request_ptr, response);
            });
    }
    virtual ~GenericNetworkTransport() = default;

private:
    virtual void do_send_request_to_server(const Request& request,
                                           util::UniqueFunction<void(const Response&)>&& cb) = 0;
};
```
(possibly with the definition of `send_request_to_server` defined in a cpp)

Then the main change sdks need to do is add `do_` to the method name and const-qualify the request parameter.

You could also keep the old name for the private virtual method and make a new name for the public method that everyone is supposed to use. A final option would be to just make both public with the same name, and overload based on the parameters that the callback takes. That only makes sense if some call sites need the request and others don't, because the ones that don't won't have to pay for keeping the request around.